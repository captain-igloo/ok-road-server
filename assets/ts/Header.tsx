import * as React from 'react';

import UserMenu from './UserMenu';

interface Props {
    showDemo?: boolean;
    showMap?: boolean;
    showRegister?: boolean;
    showSignIn?: boolean;
    user?: {
        fullName: string;
    };
}

export default function Header(props: Props) {
    const {
        showDemo,
        showMap,
        showRegister,
        showSignIn,
        user,
    } = props;

    return (
        <div className="header">
            <div className="logo">
                <a href="/">
                    <img alt="Home" src="/img/ok.svg" width="40px" height="40px" />
                </a>
                &nbsp;
                OK Road New Zealand
            </div>
            <UserMenu
                showDemo={showDemo}
                showMap={showMap}
                showRegister={showRegister}
                showSignIn={showSignIn}
                user={user}
            />
        </div>
    );
}
