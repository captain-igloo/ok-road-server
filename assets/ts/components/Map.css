body {
    margin: 0;
    padding: 0;
}

.leaflet-container {
    flex-grow: 1;
    height: 100%;
}
