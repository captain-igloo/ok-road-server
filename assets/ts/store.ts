import { combineReducers, configureStore } from '@reduxjs/toolkit';

import configReducer from './config/slice';
import friendsReducer from './friends/slice';
import mapReducer from './map/slice';
import notificationsReducer from './notifications/slice';

const rootReducer = combineReducers({
    config: configReducer,
    friends: friendsReducer,
    notifications: notificationsReducer,
    map: mapReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => configureStore({
    reducer: rootReducer,
    preloadedState,
});

export type AppStore = ReturnType<typeof setupStore>;

export type RootState = ReturnType<typeof rootReducer>;

export type AppDispatch = AppStore['dispatch'];
