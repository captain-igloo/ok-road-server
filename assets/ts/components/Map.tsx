import { DivIcon, LatLng } from 'leaflet';
import * as React from 'react';
import { Form } from 'react-bootstrap';
import { MapContainer, Marker, Polygon, Popup, TileLayer } from 'react-leaflet';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import actions from '../model/app/actions';
import { Actions, State, Dispatch } from '../model/app/types';
import { Device } from '../model/device/types';
import './Map.css';

interface Props {
    actions: Actions;
    device: any;
}

class Map extends React.PureComponent<Props> {
    public render(): React.ReactNode {
        const { actions, device } = this.props;
        return (
            <div style={{ display: 'flex', flexDirection: 'row', height: '100%' }}>
                <div style={{ width: '300px' }}>
                    <h2>Devices</h2>
                    { this.getDevices() }
                    <h2>Speed Limit</h2>
                    { device.speedLimit.description } 
                </div>
                <MapContainer center={[-40, 174]} zoom={8}>
                    <TileLayer
                        attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
                        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
                    />
                    { this.getMarkers() }
                    { /* this.getCameras() */ }
                    <Marker
                        draggable
                        eventHandlers={{
                            dragend: (e) => {
                                // console.log('drag end', e.target.getLatLng());
                                actions.setMarkerPosition(e.target.getLatLng());
                            }
                        }}
                        position={device.speedLimit.position}
                    />
                    { this.getPolygons() }
                    { /*device.polygons.map((polygon: any, index: number) => {
                        return (
                            <Polygon key={index} positions={ polygon } />
                        );
                    }) */ }
                </MapContainer>
            </div>
        );
    }

    private getPolygons(): React.ReactNode[] {
        const { device } = this.props;
        return device.speedLimit.polygons.map((positions: any, index: number) => {
            return (
                <Polygon
                    key={index}
                    positions={positions}
                />
            );
        });
        /* return [
            <Polygon
                key={99}
                positions={[new LatLng(-40, 174), new LatLng(-40, 175), new LatLng(-39, 174.5), new LatLng(-40, 174)]}
            />
        ];*/
    }
    
    private getCameras(): React.ReactNode[] {
        const { device } = this.props;
        const markers: React.ReactNode[] = [];
        Object.keys(device.cameras).forEach((key) => {
            const camera = device.cameras[key];
            markers.push(
                <Marker
                    key={camera.id}
                    position={camera.location}
                />
            );
        });
        return markers;
    }

    private getMarkers(): React.ReactNode[] {
        const { device } = this.props;
        const markers: React.ReactNode[] = [];
        Object.keys(device.devices).forEach((key) => {
            Object.keys(device.devices[key].locations).forEach((locationId) => {
                const location = device.devices[key].locations[locationId];
                const divIcon = new DivIcon({
                    html: `${location.speed} km/h`,
                });
                markers.push(
                    <Marker
                        icon={divIcon}
                        key={location.id}
                        position={location.location}
                    >
                        <Popup>asdf</Popup>
                    </Marker>,
                );
            });
        });
        /* markers.push(
            <Marker
                key="adfasdf"
                position={new LatLng(-41.231780, 174.809850)} 
            />
        );
        markers.push(
            <Marker
                key="fsefsef"
                position={new LatLng(-41.242505, 174.813068)} 
            />
        );*/
        return markers;
    }

    private getDevices(): React.ReactNode {
        const { device } = this.props;
        return Object.keys(device.devices).map((key) => {
            return (
                <Form.Group className="mb-3" controlId="formBasicCheckbox" key={key}>
                    <Form.Check
                        checked={device.devices[key].active}
                        label={device.devices[key].macAddress}
                        onChange={() => { this.onChange(device.devices[key]); }}
                        type="checkbox"
                    />
                </Form.Group>
            );
        });
    }

    private onChange(device: Device): void
    {
        const { actions } = this.props;
        actions.activateDevice(device, !device.active);
    }
}

export default connect<State, {}, any, State>((state: State): State => state, (dispatch: Dispatch) => ({
    actions: bindActionCreators(actions, dispatch),
}))(Map);
