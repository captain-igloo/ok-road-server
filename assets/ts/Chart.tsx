import {
    BarElement,
    Chart as ChartJS,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
} from 'chart.js';
import * as moment from 'moment';
import * as React from 'react';
import { Bar } from "react-chartjs-2";
import { useSelector } from 'react-redux';

import { RootState } from './store';

ChartJS.register(
    BarElement,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
);

export const options = {
    maintainAspectRatio: false,
    plugins: {
      legend: {
        display: false,
      },
    },
    responsive: true,
  };

export default function Chart() {
    const features = useSelector((state: RootState) => state.okRoad.features);

    const keys = Object.keys(features);
    const values = Array(60).fill(0);
    const labels = [];

    if (keys.length > 0) {
        const diff = features[keys[keys.length - 1]].timestamp - features[keys[0]].timestamp;
        const first = features[keys[0]].timestamp;
        const a = diff / 59;
        if (a > 0) {
            keys.forEach((key) => {
                const timestamp = features[key].timestamp;
                const index = Math.floor((timestamp - first) / a);
                values[index]++;
            });

            for (let i = first; i <= features[keys[keys.length - 1]].timestamp; i += a) {
                labels.push(moment(new Date(i * 1000)).format('h:mm a'));
            }
        }
    }

    return (
        <div className="chart-container">
            <Bar options={options} data={{
                labels,
                datasets: [{
                    backgroundColor: 'rgba(255, 99, 132, 0.5)',
                    data: values,
                }],
            }} />
        </div>
    );
}