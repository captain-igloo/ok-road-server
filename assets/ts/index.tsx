import * as Sentry from '@sentry/browser';
import * as React from 'react';
import Container from 'react-bootstrap/Container';
import * as ReactDOM from 'react-dom/client';
import { Provider } from 'react-redux';

import Header from './Header';
import Important from './Important';
import { setupStore } from './store';

const element = document.getElementById('app');
if (element) {
    const dataParams = element.getAttribute('data-params');
    let configuration;
    if (dataParams) {
        configuration = JSON.parse(dataParams);
        if (configuration.sentryDsn) {
            Sentry.init({
                dsn: configuration.sentryDsn,
            });
        }
    }
    const root = ReactDOM.createRoot(element);
    root.render(
        <Provider store={setupStore()}>
            <Header user={configuration.user} />
            <Container style={{ paddingTop: '5px' }}>
                <h1>Monitor Young Drivers</h1>
                <p>
                    Install an app on their device and keep track of:
                </p>
                <ul>
                    <li>Where they are</li>
                    <li>How fast they are driving</li>
                    <li>Their speed in relation to the speed limit</li>
                </ul>
                <Important>
                    Speed limit data is currently only available in New Zealand.
                </Important>
                <h2>Getting Started</h2>
                <ol>
                    <li>
                        Click
                        {' '}
                        <a href="/register">here</a>
                        {' '}
                        to register with OK Road.  Remember your username and password because you
                        will need it in step 3.
                    </li>
                    <li>
                        Install
                        {' '}
                        <a href="https://owntracks.org">Owntracks</a>
                        {' '}
                        on the device you wish to monitor.
                    </li>
                    <li>
                        Configure Owntracks to publish to OK Road:
                        <ul>
                            <li>Set Preferences / Connection / Endpoint / Host to "192.9.176.238".</li>
                            <li>Set Preferences / Connection / Endpoint / Port to "1883".</li>
                            <li>Set Preferences / Connection / Credentials / Username to your username.</li>
                            <li>Set Preferences / Connection / Credentials / Password to your password.</li>
                        </ul>
                    </li>
                    <li>
                        Make sure that Owntracks has permission to use location services:
                        <ul>
                            <li>Set Settings / Location / Owntracks / Location access to "Allow all the time".</li>
                            <li>Set Settings / Location / Owntracks / Use precise location to true.</li>
                        </ul>
                    </li>
                </ol>
                <img alt="Screenshot" src="/img/okroad.png" />
            </Container>
        </Provider>,
    );
}
