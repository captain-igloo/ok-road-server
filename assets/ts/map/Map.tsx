import { Icon } from 'leaflet';
import * as moment from 'moment';
import * as React from 'react';
import {
    MapContainer,
    Marker,
    Popup,
    TileLayer,
    useMap,
} from 'react-leaflet';
import { useSelector } from 'react-redux';

import { RootState } from '../store';

interface Props {
    bounds?: [[number, number], [number, number]];
}

function FitBounds(props: Props) {
    const { bounds } = props;
    const map = useMap();
    React.useEffect(() => {
        if (bounds) {
            map.fitBounds(bounds);
        }
    }, [bounds]);
    return null;
}

const formatDate = (date: Date) => {
    return moment(date).format('MMM Do YYYY h:mm:ss a');
};

export default function Map(props: Props) {
    const { bounds } = props;

    const features = useSelector((state: RootState) => state.okRoad.features);
    const markers = features.map((feature) => {
        let speedLimit;
        let image = 'gray.svg';
        let zIndexOffset = 1;
        if (feature.speedLimit !== undefined) {
            if (feature.velocity > feature.speedLimit.speedLimit) {
                image = 'red.svg';
                zIndexOffset = 3;
            } else {
                image = 'green.svg';
                zIndexOffset = 2;
            }
            speedLimit = (
                <>
                    <p>
                        <strong>Speed Limit:</strong>
                        {' '}
                        {feature.speedLimit.speedLimit}
                        km/h
                    </p>
                    <p>
                        <strong>Speed Limit Area:</strong>
                        {' '}
                        {feature.speedLimit.description}
                    </p>
                </>
            );
        }

        return (
            <Marker
                icon={new Icon({
                    iconSize: [20, 20],
                    iconUrl: `/img/${image}`,
                })}
                key={feature.id}
                position={[feature.coordinates[1], feature.coordinates[0]]}
                zIndexOffset={zIndexOffset}
            >
                <Popup>
                    <p>
                        <strong>Date:</strong>
                        {' '}
                        {formatDate(new Date(feature.timestamp * 1000))}
                    </p>
                    <p>
                        <strong>Speed:</strong>
                        {' '}
                        {feature.velocity}
                        km/h
                    </p>
                    {speedLimit}
                </Popup>
            </Marker>
        );
    });

    return (
        <MapContainer center={[-41, 174]} className="map" zoom={5} scrollWheelZoom={false}>
            <TileLayer
                attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
                url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
            />
            {markers}
            <FitBounds bounds={bounds} />
        </MapContainer>
    );
}
