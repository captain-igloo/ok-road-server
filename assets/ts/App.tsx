import * as moment from 'moment';
import * as React from 'react';
import Card from 'react-bootstrap/Card';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Form from 'react-bootstrap/Form';
import { useDispatch, useSelector } from 'react-redux';

import Chart from './Chart';
import Devices from './Devices';
import Header from './Header';
import Map from './Map';
import { setFromDate, setShowRecent, setToDate } from './slice';
import { AppDispatch, RootState } from './store';

export const useAppDispatch: () => AppDispatch = useDispatch;

export default function App() {
    const bounds = useSelector((state: RootState) => state.okRoad.bounds);
    const fromDate = useSelector((state: RootState) => state.okRoad.fromDate);
    const showRecent = useSelector((state: RootState) => state.okRoad.showRecent);
    const toDate = useSelector((state: RootState) => state.okRoad.toDate);
    
    const dispatch = useAppDispatch();

    return (
        <>
            <Header />
            <Container className="main-container" fluid>
                <Row className="main-container--row">
                    <Col className="left-container" xs={3}>
                        <Card>
                            <Card.Header>Search</Card.Header>
                            <Card.Body>
                                <Form>
                                    <Devices />
                                    <Form.Group className="mb-3">
                                        <Form.Check
                                            checked={showRecent}
                                            id="show-recent"
                                            label="Last 24 hours"
                                            onChange={(e) => {
                                                dispatch(setShowRecent(e.target.checked));
                                            }}
                                            type="checkbox"
                                        />
                                    </Form.Group>
                                    <Form.Group className="mb-3">
                                        <Form.Label>From:</Form.Label>
                                        <Form.Control
                                            disabled={showRecent}
                                            onChange={(e) => {
                                                dispatch(setFromDate((new Date(e.target.value)).getTime()));
                                            }}
                                            type="datetime-local"
                                            value={moment(new Date(fromDate)).format('Y-MM-DDTHH:mm')}
                                        />
                                    </Form.Group>
                                    <Form.Group className="mb-3">
                                        <Form.Label>To:</Form.Label>
                                        <Form.Control
                                            disabled={showRecent}
                                            onChange={(e) => {
                                                dispatch(setToDate((new Date(e.target.value)).getTime()));
                                            }}
                                            type="datetime-local"
                                            value={moment(new Date(toDate)).format('Y-MM-DDTHH:mm')}
                                        />
                                    </Form.Group>
                                    <Form.Group>
                                        <Form.Check label="Fit map" type="checkbox" />
                                    </Form.Group>
                                </Form>
                            </Card.Body>
                        </Card>
                    </Col>
                    <Col className="right-container" xs={9}>
                        <Map bounds={bounds} />
                        <Chart />
                    </Col>
                </Row>
            </Container>
        </>        
    );
}
