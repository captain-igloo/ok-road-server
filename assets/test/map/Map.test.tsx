import '../setup';

import {
    describe,
    expect,
    jest,
    test,
} from '@jest/globals';
import { render } from '@testing-library/react';
import * as React from 'react';
import { useMap } from 'react-leaflet';
import { Provider } from 'react-redux';

import Map from '../../ts/map/Map';
import { setupStore } from '../../ts/store';

jest.mock('../../ts/map/ButtonControl');

const map = {
    _controlCorners: {
        bottomright: {
            insertBefore: () => {},
        },
    },
    addLayer: () => {},
    on: () => {},
    removeLayer: () => {},
    whenReady: () => {},
};

describe('Map component', () => {
    test('Map should render properly', () => {
        (useMap as any).mockImplementation(() => map);
        const { container } = render(
            <Provider store={setupStore()}>
                <Map />
            </Provider>,
        );
        expect(container).toMatchSnapshot();
        (useMap as any).mockRestore();
    });
});
