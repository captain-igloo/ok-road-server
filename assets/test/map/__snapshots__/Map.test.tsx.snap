// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Map component Map should render properly 1`] = `
<div>
  <div
    class="map leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer gray"
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              src="https://c.tile.openstreetmap.org/5/31/19.png"
              style="width: 256px; height: 256px; left: -119px; top: -257px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              src="https://a.tile.openstreetmap.org/5/31/20.png"
              style="width: 256px; height: 256px; left: -119px; top: -1px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-disabled="false"
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            <span
              aria-hidden="true"
            >
              +
            </span>
          </a>
          <a
            aria-disabled="false"
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            <span
              aria-hidden="true"
            >
              −
            </span>
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JavaScript library for interactive maps"
          >
            <svg
              aria-hidden="true"
              class="leaflet-attribution-flag"
              height="8"
              viewBox="0 0 12 8"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h12v4H0z"
                fill="#4C7BE1"
              />
              <path
                d="M0 4h12v3H0z"
                fill="#FFD500"
              />
              <path
                d="M0 7h12v1H0z"
                fill="#E0BC00"
              />
            </svg>
             Leaflet
          </a>
           
          <span
            aria-hidden="true"
          >
            |
          </span>
           © 
          <a
            href="https://www.openstreetmap.org/copyright"
          >
            OpenStreetMap
          </a>
           contributors
        </div>
      </div>
    </div>
  </div>
</div>
`;
